kotlinx-coroutines-core/common/src/flow/operators/Context.kt

```diff
@@ -74,41 +169,163 @@ public fun <T> Flow<T>.flowOn(flowContext: CoroutineContext, bufferSize: Int = 1
-    flowContext: CoroutineContext,
-    bufferSize: Int = 16,
+    context: CoroutineContext,
     builder: Flow<T>.() -> Flow<R>
 ): Flow<R> {
-    check(flowContext, bufferSize)
-    val source = this
-    return flow {
-        /**
-         * Here we should remove a Job instance from the context.
-         * All builders are written using scoping and no global coroutines are launched, so it is safe not to provide explicit Job.
-         * It is also necessary not to mess with cancellation if multiple flowWith are used.
-         */
-        val originalContext = coroutineContext.minusKey(Job)
-        val prepared = source.flowOn(originalContext, bufferSize)
-        builder(prepared).flowOn(flowContext, bufferSize).collect { value ->
-            return@collect emit(value)
+    checkFlowContext(context)
+    return FlowWithOperator(this, context, builder)
+}
+
+// FlowWith lazily applies a builder during collection time taking the actual collectContext into account
+private class FlowWithOperator<S, T>(
+    private val flow: Flow<S>,
+    private val context: CoroutineContext,
+    private val builder: Flow<S>.() -> Flow<T>
+) : Flow<T> {
+    fun buildFlow(collectContext: CoroutineContext): Flow<T> {
+        // we only restore context elements that are explicitly specified in the context
+        val restoredContext = context.fold<CoroutineContext>(EmptyCoroutineContext) { acc, element ->
+            val restoredElement = collectContext[element.key]
+            if (restoredElement == null) acc else acc + restoredElement
         }
+        // todo: what if the element was not present in the context? shall we remove it?
+        return flow.flowOn(restoredContext).builder().flowOn(context)
     }
+
+    override suspend fun collect(collector: FlowCollector<T>) =
+        buildFlow(coroutineContext).collect(collector)
+
+    // debug toString
+    override fun toString(): String =
+        "$flow -> flowWith[context=$context, builder=$builder]"
 }
 
-private fun check(flowContext: CoroutineContext, bufferSize: Int) {
-    require(flowContext[Job] == null) {
-        "Flow context cannot contain job in it. Had $flowContext"
+// ChannelFlow implementation that operates on another flow before it
+internal abstract class ChannelFlowOperator<S, T>(
+    @JvmField val flow: Flow<S>,
+    context: CoroutineContext,
+    capacity: Int
+) : ChannelFlow<T>(context, capacity) {
+    protected abstract suspend fun flowCollect(collector: FlowCollector<T>)
+
+    // Changes collecting context upstream to the specified newContext, while collecting in the original context
+    private suspend fun collectWithContextUndispatched(collector: FlowCollector<T>, newContext: CoroutineContext) {
+        val originalContextCollector = collector.withUndispatchedContextCollector(coroutineContext)
+        // invoke flowCollect(originalContextCollector) in the newContext
+        val flowCollectRef: suspend (FlowCollector<T>) -> Unit = { flowCollect(it) }
```

Comment:
`flowCollectRef` can be inlined into use-site

Reply:
Method references to not work on current version of JS. Left TODO there.